flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/filemapping/FileMappingManager.java

```diff
@@ -0,0 +1,252 @@
+                dstEntry.release();
+            }
+            mappingTable.remove(src);
+            mappingTable.put(dst, srcEntry);
+        } else { // rename directory
+            // 1. rename file in old directory
+            List<String> toRename = new ArrayList<>();
+            for (String key : mappingTable.keySet()) {
+                if (isParentDir(key, src) || key.equals(src)) {
+                    toRename.add(key);
+                }
+            }
+            if (!toRename.isEmpty()) {
+                for (String key : toRename) {
+                    MappingEntry sourceEntry = mappingTable.remove(key);
+                    String renamedDst = key.replace(src, dst);
+                    LOG.trace("rename: {} -> {}", key, renamedDst);
+                    sourceEntry.sourcePath = sourceEntry.sourcePath.replace(src, dst);
+                    mappingTable.put(renamedDst, sourceEntry);
+                }
+            }
+            // 2. rename directory in file system physically
+            Path srcPath = new Path(src);
+            Path dstPath = new Path(dst);
+
+            // The rename is not atomic for ForSt. Some FileSystems e.g. HDFS, OSS does not allow a
+            // renaming if the target already exists. So, we delete the target before attempting the
+            // rename.
+            if (fileSystem.exists(dstPath)) {
+                boolean deleted = fileSystem.delete(dstPath, true);
+                if (!deleted) {
+                    throw new IOException("Fail to rename path, src:" + src + ", dst:" + dst);
+                }
+            }
+            fileSystem.rename(srcPath, dstPath);
+        }
+        return true;
+    }
+
+    /**
+     * Delete a file or directory from mapping table and file system, the directory deletion may be
+     * deferred.
+     *
+     * @param file to be deleted
+     * @param recursive whether to delete recursively
+     * @return true if the file or directory is deleted successfully, false otherwise.
+     * @throws IOException
+     */
+    public boolean deleteFile(Path file, boolean recursive) throws IOException {
+        String fileStr = file.toString();
+        MappingEntry entry = mappingTable.getOrDefault(fileStr, null);
+        LOG.trace("delete: {}, source:{}", file, entry == null ? "null" : entry.sourcePath);
+        if (entry != null) { // delete file
+            mappingTable.remove(fileStr);
+            if (!entry.isDirectory) {
+                entry.release();
+                return true;
+            }
+        }
+
+        // delete directory
+        MappingEntry parentEntry = new MappingEntry(0, fileSystem, fileStr, true);
+
+        boolean matched = false;
+        for (MappingEntry mappingEntry : mappingTable.values()) {
+            if (isParentDir(mappingEntry.sourcePath, fileStr)) {
+                parentEntry.retain();
+                mappingEntry.parentDir = parentEntry;
```

Comment:
Ah...  Shall we create all the `MappingEntry` all the way to the base directory? or the input `file`

Reply:
If the new directory is the parent of old `parent dir`, the old `parent dir` is replaced by base dir now.